integrations/optimizely/lib/index.js

```diff
@@ -426,205 +264,85 @@ Optimizely.prototype.setEffectiveReferrer = function(referrer) {
-          experiment: {
-            id: currentExperimentId,
-            name: data.experiments[currentExperimentId].name
-          },
-          variations: [],
-          /** Segment added code */
-          // we need to send sectionName for multivariate experiments
-          sections: data.sections
-          /**/
-        };
-
-        /** Segment added code */
-        // for backward compatability since we send referrer with the experiment properties
-        if (
-          state.redirectExperiment &&
-          currentExperimentId === state.redirectExperiment.experimentId &&
-          state.redirectExperiment.referrer
-        ) {
-          activeExperimentState.experiment.referrer =
-            state.redirectExperiment.referrer;
-        }
-        /**/
-
-        var variationIds =
-          state.variationIdsMap[activeExperimentState.experiment.id];
-        for (var j = 0; j < variationIds.length; j++) {
-          var id = variationIds[j];
-          var name = data.variations[id].name;
-          activeExperimentState.variations.push({
-            id: id,
-            name: name
-          });
-        }
-        sendExperimentData(activeExperimentState);
+      if (referrer) {
+        self.setEffectiveReferrer(referrer);
+        return referrer; // Segment added code: so I can pass this referrer value in cb
       }
     }
   };
 
   /**
-   * This function fetches all the campaign data from the new Optimizely client
-   * and calls the functions provided in the arguments with the data that needs to
-   * be used for sending information. It is recommended to leave this function as is
-   * and to create your own implementation of the functions referrerOverride and
-   * sendCampaignData.
-   *
-   * @param {Function} referrerOverride - This function is called if the effective referrer value differs from
-   *   the current document.referrer value. The only argument provided is the effective referrer value.
-   * @param {Function} sendCampaignData - This function is called for every running campaign on the page.
-   *   The function is called with the campaignState for the activated campaign
+   * A campaign or experiment can be activated after we have initialized.
+   * This function registers a listener that listens to newly activated campaigns and
+   * handles them.
    */
-  var initNewOptimizelyIntegration = function(
-    referrerOverride,
-    sendCampaignData
-  ) {
-    var newActiveCampaign = function(id, referrer) {
-      var state = window.optimizely.get && window.optimizely.get('state');
-      if (state) {
-        var activeCampaigns = state.getCampaignStates({
-          isActive: true
-        });
-        var campaignState = activeCampaigns[id];
-        // Segment added code: in case this is a redirect experiment
-        if (referrer) campaignState.experiment.referrer = referrer;
-        sendCampaignData(campaignState);
+  var registerFutureActiveCampaigns = function() {
```

Comment:
we talked a little bit about potentially bringing these out as private functions? It would make this suite much easier to test IMO. What are your thoughts?

If not, it feels a little questionable to declare these as functions and not really re-use them. Maybe we can unfurl them and supplement with comments? WDYT?

Reply:
Fair points! Of the three remaining nested functions (see the latest commit) I think some are going to be easily reusable and some aren't.

Maybe can extract the reusable ones?

What would we do with remaining ones - inline them? One could argue that single-use nested functions are helpful for documenting what each block of code does, but I think code comments might suffice too.